tfhe/src/integer/ciphertext/compact_list.rs

```diff
@@ -46,44 +384,52 @@ impl CompactCiphertextList {
     /// Panics if the constituents are not compatible with each others.
     pub fn from_raw_parts(
         ct_list: crate::shortint::ciphertext::CompactCiphertextList,
-        num_blocks_per_integer: usize,
+        info: Vec<DataKind>,
     ) -> Self {
         assert_eq!(
-            ct_list.ct_list.lwe_ciphertext_count().0 % num_blocks_per_integer,
-            0,
+            ct_list.ct_list.lwe_ciphertext_count().0,
+            info.len(),
             "CompactCiphertextList LweCiphertextCount is expected \
-            to be a multiple of {num_blocks_per_integer}, got {:?}",
+            to be equal to the info vec {} vs {:?}",
+            info.len(),
             ct_list.ct_list.lwe_ciphertext_count()
         );
 
-        Self {
-            ct_list,
-            num_blocks_per_integer,
-        }
+        Self { ct_list, info }
     }
 
     pub fn ciphertext_count(&self) -> usize {
-        self.ct_list.ct_list.lwe_ciphertext_count().0 / self.num_blocks_per_integer
+        self.info.len()
     }
 
-    pub fn expand<T: IntegerRadixCiphertext>(&self) -> Vec<T> {
-        let mut all_block_iter = self.ct_list.expand().into_iter();
-        let num_ct = self.ciphertext_count();
-        let mut ciphertexts = Vec::with_capacity(num_ct);
+    pub fn expand(&self, sks: &ServerKey) -> crate::Result<CompactCiphertextListExpander> {
+        let expanded_blocks = self.ct_list.expand();
 
-        for _ in 0..num_ct {
-            let ct_blocks = all_block_iter
-                .by_ref()
-                .take(self.num_blocks_per_integer)
-                .collect::<Vec<_>>();
-            if ct_blocks.len() < self.num_blocks_per_integer {
-                break;
-            }
-            let ct = T::from(ct_blocks);
-            ciphertexts.push(ct);
+        let mut conformance_params = sks.key.conformance_params();
+        conformance_params.degree = self.ct_list.degree;
+        if !self.is_conformant_with_shortint_params(conformance_params) {
+            return Err(crate::Error::new(
+                "This compact list is not conformant with the given server key".to_string(),
+            ));
         }
 
-        ciphertexts
+        let expanded_blocks = if self.ct_list.degree.get() > self.ct_list.message_modulus.0 - 1 {
+            extract_message_and_carries(expanded_blocks, sks)
+        } else {
+            expanded_blocks
+        };
+
+        Ok(CompactCiphertextListExpander::new(
+            expanded_blocks,
+            self.info.clone(),
+        ))
+    }
+
+    /// This simply expands the list without splitting messages if they were packed
+    /// caller should check using [Self::is_packed] to know if data is packed
+    pub(crate) fn expand_without_key(&self) -> CompactCiphertextListExpander {
```

Comment:
expand_without_packing ? not sure the name is better than this one

Reply:
`expand_without_unpacking` you mean ? It might be better yes

Reply:
well without unpacking reads a bit weird as it could mean "it is packed, but I don't unpack" not sure how to name those things 🙄

Reply:
The name `expand_without_unpacking` seems correct to me because this function will expand the data but never unpack (ie split 1 block into 2), this means that if the blocks are packed this function should not be called, the caller shall use `.is_packed()` before

Reply:
fair enough